tracepoint_decode/src/charconv.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+use core::fmt;
+use core::str;
+
+use crate::filters;
+
+/// Returns true if `maybe_valid_ch32` is a valid Unicode scalar value.
+/// Returns false for surrogate code points (0xD800..0xE000) and values >= 0x110000.
+#[inline]
+const fn is_valid_char(maybe_valid_ch32: u32) -> bool {
+    // This is an optimized check for !0xD800..0xE000 && < 0x110000.
+    return (maybe_valid_ch32 ^ 0xD800).wrapping_sub(0x800) < 0x110000 - 0x800;
+}
+
+#[inline]
+fn char_from_validated_u32(valid_ch32: u32) -> char {
+    debug_assert!(!(0xD800..=0xDFFF).contains(&valid_ch32) && valid_ch32 < 0x110000);
+    return unsafe { char::from_u32_unchecked(valid_ch32) };
+}
+
+#[inline]
+fn str_from_validated_utf8(valid_utf8: &[u8]) -> &str {
+    debug_assert!(core::str::from_utf8(valid_utf8).is_ok());
+    return unsafe { str::from_utf8_unchecked(valid_utf8) };
+}
+
+/// If ch32 is valid, returns the char. Otherwise, returns the replacement character.
+pub fn char_from_u32(ch32: u32) -> char {
```

Comment:
```Rust
pub fn char_from_u32 (ch32: u32)_-> char {
    if let Some(c) = char::from_u32(ch32) { c } else { char::REPLACEMENT_CHARACTER }
}

Reply:
No more rolling our own version of built-in functions. Everybody wins!

Reply:
```Rust
pub fn char_from_u32 (ch32: u32)_-> char {
    char::from_u32(ch32).unwrap_or( char::REPLACEMENT_CHARACTER )
}
```

Reply:
(it's fine to not put return everywhere. it's the rust way)

Reply:
I like your implementation.

I started trying "leave return out" and I made some bugs and decided that I'm just not ready yet. :P